sdk/azidentity/azure_cli_credential_client.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+const (
+	DefaultSuffix        = "/.default"
+	AzureCLINotInstalled = "Azure CLI not installed"
+	AzNotLogIn           = "ERROR: Please run 'az login'"
+	WinAzureCLIError     = "'az' is not recognized"
+	invalidResourceError = "Resource is not in expected format. Only alphanumeric characters, '.', ':', '-', and '/' are allowed"
+)
+
+type azureCLIAccessToken struct {
+	Token        string `json:"accessToken"`
+	ExpiresOn    string `json:"expiresOn"`
+	Subscription string `json:"subscription"`
+	Tenant       string `json:"tenant"`
+	TokenType    string `json:"tokenType"`
+}
+
+type shellClient interface {
+	getAzureCLIAccessToken(command string) ([]byte, string, error)
+}
+
+// AzureCLICredentialClient provides the base for authenticating with Azure CLI Credential.
+type azureCLICredentialClient struct {
+}
+
+// NewAzureCLICredentialClient creates a new instance of the AzureCLICredentialClient.
+func newAzureCLICredentialClient() *azureCLICredentialClient {
+	return &azureCLICredentialClient{}
+}
+
+// Authenticate runs a Azure CLI command for a Azure CLI Credential and returns the resulting Access Token or
+// an error in case of authentication failure.
+// ctx: The current request context
+// scopes: The scopes required for the token
+func (c *azureCLICredentialClient) authenticate(ctx context.Context, scopes []string, credentialClient shellClient) (*azcore.AccessToken, error) {
+	// The command that Azure CLI would be run
+	command := "az account get-access-token --output json"
+	if scopes != nil {
+
+		// covert the scopes to a resource string
+		resource := c.scopeToResource(scopes)
+
+		// Validate the resource to make sure it doesn't include shell-meta characters since it gets sent as a command line argument to Azure CLI
+		isResourceMatch, error := regexp.MatchString("^[0-9a-zA-Z-.:/]+$", resource)
+		if error != nil {
+			return nil, error
+		}
+		if !isResourceMatch {
+			return nil, fmt.Errorf(invalidResourceError)
+		}
+
+		command = command + " --resource " + resource
+	}
+	// Execute Azure CLI command(az account get-access-token --output json --resource) to get a token to authenticate
+	out, errout, err := credentialClient.getAzureCLIAccessToken(command)
+
+	// Determining Azure CLI errors
+	if err != nil {
+		isLoginError := strings.HasPrefix(errout, AzNotLogIn)
```

Comment:
Normally in Go, if an error is not nil, then the other returned value should not be examined. Here, if err is not nil (an error occurred), then you ARE looking at errout. This doesn't feel right to me. Although it may be but there MUST be some comments explaining how this should be used if it goes against the normal Go way of doing things.

Also, it is very uncommon in Go to have a method return more than 2 values. Maybe there is a better way of accomplishing what you're trying to do here?

Reply:
Thanks. 
Return this error as CredentialUnavaiableError driectly.